File: src/lib/services/catalog-service.ts
import type { Database } from '../db-types';
import { ResourceNotFoundError } from '../errors';
import type { Repositories } from '../repositories';
import type { Event, Ticket } from '../types';

// Não existia no projeto original (só havia API de reservas e pagamentos); é o mínimo que o
// front-end precisa para listar eventos e desenhar o mapa de assentos.
export interface CatalogService {
  listAvailableEvents(): Promise<Event[]>;
  getEvent(id: number): Promise<Event>;
  listSeats(eventId: number): Promise<Ticket[]>;
}

export function createCatalogService(deps: { db: Database; repos: Repositories }): CatalogService {
  const { db, repos } = deps;

  async function getEvent(id: number): Promise<Event> {
    const event = await repos.events.findById(db.pool, id);
    if (!event) throw new ResourceNotFoundError('Event not found');
    return event;
  }

  return {
    listAvailableEvents: () => repos.events.findAvailable(db.pool),
    getEvent,
    async listSeats(eventId) {
      await getEvent(eventId);
      return repos.tickets.listByEvent(db.pool, eventId);
    },
  };
}

File: backend/app/services/catalog_service.py
"""Não existia no projeto original (só havia API de reservas e pagamentos); é o mínimo que o
front-end precisa para listar eventos e desenhar o mapa de assentos. Equivalente a
services/catalog-service.ts."""
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker

from app.errors import ResourceNotFoundError
from app.models import Event, Ticket
from app.repositories import events as events_repo
from app.repositories import tickets as tickets_repo


class CatalogService:
    def __init__(self, session_factory: async_sessionmaker[AsyncSession]) -> None:
        self._session_factory = session_factory

    async def list_available_events(self) -> list[Event]:
        async with self._session_factory() as session:
            return await events_repo.find_available(session)

    async def get_event(self, event_id: int) -> Event:
        async with self._session_factory() as session:
            event = await events_repo.find_by_id(session, event_id)
        if event is None:
            raise ResourceNotFoundError("Event not found")
        return event

    async def list_seats(self, event_id: int) -> list[Ticket]:
        await self.get_event(event_id)
        async with self._session_factory() as session:
            return await tickets_repo.list_by_event(session, event_id)
